File: cms_detector.py
#!/usr/bin/env python3
# pylint: disable=line-too-long, broad-exception-caught
""" Detect CMS/frameworks using lightweight HTTP heuristics."""

import argparse
import base64
import binascii
import json
import os
import re
import sys
import urllib.parse

import requests
import urllib3

# ANSI colors (preserved from your original)
PURPLE = '\033[95m'
CYAN = '\033[96m'
DARKCYAN = '\033[36m'
BLUE = '\033[94m'
GREEN = '\033[92m'
YELLOW = '\033[93m'
RED = '\033[91m'
BOLD = '\033[1m'
UNDERLINE = '\033[4m'
END = '\033[0m'


class FingerprintMatcher:
    """
    Prepares and evaluates fingerprints efficiently.
    Compiles regexes and normalizes values for faster matching.
    """
    def __init__(self, fingerprints):
        # Pre-compile regex patterns and keep structure as-is for other types
        self.cms_list = []
        for cms in fingerprints:
            prepared = []
            for fp in cms.get("fingerprints", []):
                fptype = fp.get("type")
                if fptype == "regex":
                    pattern = fp.get("value", "")
                    try:
                        compiled = re.compile(pattern, re.IGNORECASE | re.DOTALL)
                    except re.error:
                        continue
                    prepared.append({"type": "regex", "compiled": compiled, "raw": pattern})
                else:
                    prepared.append(fp)
            self.cms_list.append({"name": cms.get("name", "Unknown"), "fingerprints": prepared})

    def match(self, response: requests.Response, response_text_lower: str):
        """
        Returns list of dicts: {"name": <cms>, "matched_by": [<description>...]}
        """
        results = []

        headers = response.headers
        cookies = response.cookies

        for cms in self.cms_list:
            matched_by = []
            for frpr in cms["fingerprints"]:
                t = frpr.get("type")

                # Body regex
                if t == "regex":
                    if frpr["compiled"].search(response_text_lower):
                        matched_by.append(f"regex:{frpr['raw']}")

                # Body contains
                elif t == "string_contains":
                    value = str(frpr.get("value", "")).lower()
                    if value and value in response_text_lower:
                        matched_by.append(f"string_contains:{frpr['value']}")

                elif t == "strings_contain":
                    required_values = str(frpr.get("value", "")).lower().split("|")
                    if required_values and all(v in response_text_lower for v in required_values):
                        matched_by.append(f"strings_contain:{frpr['value']}")

                # Header checks
                elif t == "header_key_equals":
                    key = frpr.get("value", "")
                    if key and key in headers:
                        matched_by.append(f"header_key_equals:{key}")

                elif t == "header_key_value":
                    key = frpr.get("key", "")
                    value = str(frpr.get("value", "")).lower()
                    if key in headers and str(headers.get(key, "")).lower() == value:
                        matched_by.append(f"header_key_value:{key}={frpr['value']}")

                elif t == "header_key_value_contains":
                    key = frpr.get("key", "")
                    value = str(frpr.get("value", "")).lower()
                    if key in headers and value in str(headers.get(key, "")).lower():
                        matched_by.append(f"header_key_value_contains:{key}~{frpr['value']}")

                # Cookie checks
                elif t == "cookie_key_equals":
                    key = frpr.get("value", "")
                    if key and key in cookies:
                        matched_by.append(f"cookie_key_equals:{key}")

                elif t == "cookie_key_value":
                    key = frpr.get("key", "")
                    value = str(frpr.get("value", "")).lower()
                    if key in cookies and str(cookies.get(key, "")).lower() == value:
                        matched_by.append(f"cookie_key_value:{key}={frpr['value']}")

                elif t == "cookie_key_value_contains":
                    key = frpr.get("key", "")
                    value = str(frpr.get("value", "")).lower()
                    if key in cookies and value in str(cookies.get(key, "")).lower():
                        matched_by.append(f"cookie_key_value_contains:{key}~{frpr['value']}")

                elif t == "cookie_key_value_b64_json_keys":
                    key = frpr.get("key", "")
                    required_keys = str(frpr.get("value", "")).lower().split("|")
                    if key in cookies:
                        try:
                            url_decoded = urllib.parse.unquote(cookies.get(key, ""))
                            b64_decoded = base64.b64decode(url_decoded)
                            json_decoded = json.loads(b64_decoded)

                            json_keys_lower = {str(k).lower() for k in json_decoded.keys()}
                            if all(k in json_keys_lower for k in required_keys):
                                matched_by.append(f"cookie_key_value_b64_json_keys:{key} has {frpr['value']}")
                        except (binascii.Error, json.decoder.JSONDecodeError, ValueError, TypeError):
                            pass

                elif t == "cookie_substr_key_value_b64_type":
                    end_len = int(frpr.get("length", 0))
                    key_suffix = frpr.get("key", "")
                    expected_type_name = frpr.get("value", "")
                    try:
                        for cookie in cookies:
                            if cookie.name[end_len:] == key_suffix:
                                try:
                                    url_decoded = urllib.parse.unquote(cookie.value)
                                    b64_decoded = base64.b64decode(url_decoded)
                                    if type(b64_decoded).__name__ == expected_type_name:
                                        matched_by.append(
                                            f"cookie_substr_key_value_b64_type:*{key_suffix} -> {expected_type_name}"
                                        )
                                        break
                                except binascii.Error:
                                    continue
                    except Exception:
                        pass

            if matched_by:
                results.append({"name": cms["name"], "matched_by": matched_by})

        return results


class CMSDetector:
    """ CMS Detector """

    def __init__(self, host: str, *, raw: bool = False, json_out: bool = False, insecure: bool = False, timeout: int = 10, user_agent: str | None = None):
        self.session = requests.Session()
        self.host = self._normalize_host(host)
        self.raw = raw
        self.json_out = json_out
        self.insecure = insecure
        self.timeout = timeout
        self.user_agent = user_agent or "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36"

        self._headers = {
            "User-Agent": self.user_agent,
            "Cache-Control": "max-age=0",
            "sec-ch-ua": "\"Chromium\";v=\"126\", \"Google Chrome\";v=\"126\", \"Not;A=Brand\";v=\"99\"",
            "sec-ch-ua-mobile": "?0",
            "sec-ch-ua-platform": "\"macOS\"",
            "sec-fetch-dest": "document",
            "sec-fetch-mode": "navigate",
            "sec-fetch-site": "none",
            "upgrade-insecure-requests": "1",
            "referer": self.host
        }

        self.fingerprints = self._load_fingerprints()
        self.matcher = FingerprintMatcher(self.fingerprints)


    @staticmethod
    def _normalize_host(host: str) -> str:
        host = (host or "").strip()
        if not host.startswith(("http://", "https://")):
            host = f"http://{host}"
        # remove trailing slash
        return host.rstrip("/")


    @staticmethod
    def _load_fingerprints():
        fp_path = os.path.join(".", "fingerprints.json")
        if os.path.exists(fp_path):
            with open(fp_path, "r", encoding="utf-8") as f:
                return json.load(f)
        print(f'{RED}[!] Could not find "fingerprints.json".{END}')
        sys.exit(1)


    def scan(self):
        """ Scan host for CMS """
        if not (self.raw or self.json_out):
            print(PURPLE + BOLD + "CMS Detector" + END)
            print(PURPLE + "[•] Made by: https://github.com/joshuavanderpoll/CMS-Detector" + END)
            print(f"{BLUE}[@] Scanning host {DARKCYAN}\"{self.host}\"{BLUE}...{END}")

        try:
            resp = self.session.get(
                self.host,
                headers=self._headers,
                allow_redirects=True,
                timeout=self.timeout,
                verify=not self.insecure,
            )
        except requests.exceptions.RequestException as e:
            msg = f"{RED}[!] Could not retrieve host. Error: {e}{END}"
            if self.json_out:
                print(json.dumps({"host": self.host, "error": str(e)}, ensure_ascii=False))
            else:
                print(msg)
            sys.exit(1)

        body_lower = resp.text.lower().replace("'", "\"").strip()
        matches = self.matcher.match(resp, body_lower)
        detected = bool(matches)

        if self.json_out:
            out = {
                "host": self.host,
                "status_code": resp.status_code,
                "detected": detected,
                "matches": matches,
                "timing_ms": int(resp.elapsed.total_seconds() * 1000) if resp.elapsed else None,
                "redirects": len(resp.history),
            }
            print(json.dumps(out, ensure_ascii=False))
        elif self.raw:
            if detected:
                # print all matched
                for m in matches:
                    print(GREEN + m["name"].lower().replace(" ", "_") + END)
            else:
                print("null")
        else:
            if detected:
                names = ", ".join(f"\"{m['name']}\"" for m in matches)
                print(f"{GREEN}[√] \"{self.host}\" is using {BLUE}{names}{GREEN}!{END}")
                for m in matches:
                    # Show a concise reason for each CMS
                    reasons = "; ".join(m["matched_by"][:3])
                    if len(m["matched_by"]) > 3:
                        reasons += f" (+{len(m['matched_by'])-3} more)"
                    print(f"{CYAN}    ↳ matched by: {reasons}{END}")
            else:
                print(f"{YELLOW}[!] No CMS could be detected.{END}")

        # Exit codes: 0=found, 2=no match
        sys.exit(0 if detected else 2)


def main():
    """ CLI entry point """
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    parser = argparse.ArgumentParser(
        prog="CMS Detector",
        description="Detect common CMS/frameworks using lightweight HTTP heuristics.",
        epilog="Created by: https://github.com/joshuavanderpoll"
    )
    parser.add_argument('--host', type=str, help="Host or URL to scan (e.g., example.com or https://example.com)")
    parser.add_argument('--raw', default=False, action='store_true', help="Print only result name(s) in lowercase with underscores; print 'null' on no match.")
    parser.add_argument('--json', dest='json_out', default=False, action='store_true', help="Output a structured JSON result.")
    parser.add_argument('--timeout', type=int, default=10, help="HTTP timeout in seconds (default: 10).")
    parser.add_argument('--insecure', default=False, action='store_true', help="Skip TLS verification (verify=False).")
    parser.add_argument('--ua', dest='user_agent', type=str, default=None, help="Custom User-Agent string.")

    args = parser.parse_args()

    # Interactive prompt fallback (preserved)
    host = args.host
    if host is None and not args.json_out and not args.raw:
        host = input(PURPLE + "[?] Enter host to scan : " + END)

    elif host is None:
        # If --json or --raw and no host provided, fail cleanly
        print(json.dumps({"error": "Missing --host"}, ensure_ascii=False) if args.json_out
              else f"{RED}[!] Missing --host{END}")
        sys.exit(1)

    detector = CMSDetector(
        host=host,
        raw=args.raw,
        json_out=args.json_out,
        insecure=args.insecure,
        timeout=args.timeout,
        user_agent=args.user_agent
    )
    detector.scan()


if __name__ == "__main__":
    main()

File: cms_detector.go
package main

import (
	"bufio"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/joshuavanderpoll/CMS-Detector/fingerprints"
)

const (
	GREEN    = "\033[92m"
	BLUE     = "\033[94m"
	PURPLE   = "\033[95m"
	CYAN     = "\033[96m"
	DARKCYAN = "\033[36m"
	YELLOW   = "\033[93m"
	RED      = "\033[91m"
	BOLD     = "\033[1m"
	END      = "\033[0m"
)

type Fingerprint struct {
	Type   string `json:"type"`
	Value  string `json:"value,omitempty"`
	Key    string `json:"key,omitempty"`
	Length int    `json:"length,omitempty"`
}

type CMS struct {
	Name         string        `json:"name"`
	Fingerprints []Fingerprint `json:"fingerprints"`
}

type preparedFP struct {
	fp    Fingerprint
	re    *regexp.Regexp // compiled when type == "regex"
	rawRe string
}

type preparedCMS struct {
	Name string
	FPs  []preparedFP
}

type matchResult struct {
	Name      string   `json:"name"`
	MatchedBy []string `json:"matched_by"`
}

type jsonOut struct {
	Host       string        `json:"host"`
	StatusCode int           `json:"status_code"`
	Detected   bool          `json:"detected"`
	Matches    []matchResult `json:"matches"`
	TimingMS   int64         `json:"timing_ms"`
	Redirects  int           `json:"redirects"`
	Error      string        `json:"error,omitempty"`
}

func normalizeHost(h string) string {
	h = strings.TrimSpace(h)
	if h == "" {
		return h
	}
	if !(strings.HasPrefix(h, "http://") || strings.HasPrefix(h, "https://")) {
		h = "http://" + h
	}
	return strings.TrimRight(h, "/")
}

func readFingerprints() ([]CMS, error) {
	var arr []CMS
	dec := json.NewDecoder(strings.NewReader(string(fingerprints.Data)))
	if err := dec.Decode(&arr); err != nil {
		return nil, err
	}
	return arr, nil
}

func prepare(cms []CMS) []preparedCMS {
	out := make([]preparedCMS, 0, len(cms))
	for _, c := range cms {
		pc := preparedCMS{Name: c.Name}
		for _, fp := range c.Fingerprints {
			if fp.Type == "regex" {
				pat := "(?is)" + fp.Value
				re, err := regexp.Compile(pat)
				if err != nil {
					// skip invalid regex but continue
					continue
				}
				pc.FPs = append(pc.FPs, preparedFP{fp: fp, re: re, rawRe: fp.Value})
			} else {
				pc.FPs = append(pc.FPs, preparedFP{fp: fp})
			}
		}
		out = append(out, pc)
	}
	return out
}

func headerHasKey(h http.Header, key string) bool {
	return h.Get(key) != ""
}

func headerEquals(h http.Header, key, val string) bool {
	return strings.EqualFold(h.Get(key), val)
}

func headerContains(h http.Header, key, sub string) bool {
	return strings.Contains(strings.ToLower(h.Get(key)), strings.ToLower(sub))
}

func cookiesByName(cookies []*http.Cookie, name string) *http.Cookie {
	for _, c := range cookies {
		if c.Name == name {
			return c
		}
	}
	return nil
}

func cookieExists(cookies []*http.Cookie, name string) bool {
	return cookiesByName(cookies, name) != nil
}

func cookieEquals(cookies []*http.Cookie, name, val string) bool {
	c := cookiesByName(cookies, name)
	if c == nil {
		return false
	}
	return strings.EqualFold(c.Value, val)
}

func cookieContains(cookies []*http.Cookie, name, sub string) bool {
	c := cookiesByName(cookies, name)
	if c == nil {
		return false
	}
	return strings.Contains(strings.ToLower(c.Value), strings.ToLower(sub))
}

func decodeBase64Any(s string) ([]byte, error) {
	u, _ := url.QueryUnescape(s)

	// try StdEncoding with padding fix
	normalize := func(x string) string {
		if m := len(x) % 4; m != 0 {
			x += strings.Repeat("=", 4-m)
		}
		return x
	}
	if b, err := base64.StdEncoding.DecodeString(normalize(u)); err == nil {
		return b, nil
	}
	// try URL encoding
	if b, err := base64.URLEncoding.DecodeString(normalize(u)); err == nil {
		return b, nil
	}
	// raw std
	if b, err := base64.RawStdEncoding.DecodeString(u); err == nil {
		return b, nil
	}
	// raw url
	if b, err := base64.RawURLEncoding.DecodeString(u); err == nil {
		return b, nil
	}
	return nil, fmt.Errorf("base64 decode failed")
}

func match(prepared []preparedCMS, resp *http.Response, bodyLower string) []matchResult {
	results := []matchResult{}
	h := resp.Header
	cookies := resp.Cookies()

	for _, c := range prepared {
		var matchedBy []string
		for _, pf := range c.FPs {
			fp := pf.fp
			switch fp.Type {

			case "regex":
				if pf.re != nil && pf.re.FindStringIndex(bodyLower) != nil {
					matchedBy = append(matchedBy, "regex:"+pf.rawRe)
				}

			case "string_contains":
				val := strings.ToLower(fp.Value)
				if val != "" && strings.Contains(bodyLower, val) {
					matchedBy = append(matchedBy, "string_contains:"+fp.Value)
				}

			case "strings_contain":
				parts := strings.Split(strings.ToLower(fp.Value), "|")
				all := true
				for _, p := range parts {
					if !strings.Contains(bodyLower, p) {
						all = false
						break
					}
				}
				if len(parts) > 0 && all {
					matchedBy = append(matchedBy, "strings_contain:"+fp.Value)
				}

			case "header_key_equals":
				key := fp.Value
				if key != "" && headerHasKey(h, key) {
					matchedBy = append(matchedBy, "header_key_equals:"+key)
				}

			case "header_key_value":
				key := fp.Key
				val := fp.Value
				if key != "" && headerEquals(h, key, val) {
					matchedBy = append(matchedBy, fmt.Sprintf("header_key_value:%s=%s", key, val))
				}

			case "header_key_value_contains":
				key := fp.Key
				val := fp.Value
				if key != "" && headerContains(h, key, val) {
					matchedBy = append(matchedBy, fmt.Sprintf("header_key_value_contains:%s~%s", key, val))
				}

			case "cookie_key_equals":
				name := fp.Value
				if name != "" && cookieExists(cookies, name) {
					matchedBy = append(matchedBy, "cookie_key_equals:"+name)
				}

			case "cookie_key_value":
				name := fp.Key
				val := strings.ToLower(fp.Value)
				if name != "" && cookieEquals(cookies, name, val) { // case-insensitive equality
					matchedBy = append(matchedBy, fmt.Sprintf("cookie_key_value:%s=%s", name, fp.Value))
				} else {
					// equality with case-sensitive value (fallback)
					c := cookiesByName(cookies, name)
					if c != nil && c.Value == fp.Value {
						matchedBy = append(matchedBy, fmt.Sprintf("cookie_key_value:%s=%s", name, fp.Value))
					}
				}

			case "cookie_key_value_contains":
				name := fp.Key
				val := fp.Value
				if name != "" && cookieContains(cookies, name, val) {
					matchedBy = append(matchedBy, fmt.Sprintf("cookie_key_value_contains:%s~%s", name, val))
				}

			case "cookie_key_value_b64_json_keys":
				name := fp.Key
				if c := cookiesByName(cookies, name); c != nil {
					if b, err := decodeBase64Any(c.Value); err == nil {
						var obj map[string]any
						if json.Unmarshal(b, &obj) == nil {
							req := strings.Split(strings.ToLower(fp.Value), "|")
							have := map[string]bool{}
							for k := range obj {
								have[strings.ToLower(k)] = true
							}
							ok := true
							for _, k := range req {
								if !have[k] {
									ok = false
									break
								}
							}
							if ok {
								matchedBy = append(matchedBy, fmt.Sprintf("cookie_key_value_b64_json_keys:%s has %s", name, fp.Value))
							}
						}
					}
				}

			case "cookie_substr_key_value_b64_type":
				l := fp.Length
				keySuffix := fp.Key
				expect := fp.Value
				for _, c := range cookies {
					name := c.Name
					idx := len(name) + l // l can be negative
					if idx < 0 {
						idx = 0
					}
					if idx > len(name) {
						idx = len(name)
					}
					suffix := name[idx:]
					if suffix == keySuffix {
						if b, err := decodeBase64Any(c.Value); err == nil {
							switch strings.ToLower(expect) {
							case "bytes":
								// any successful decode counts
								matchedBy = append(matchedBy, fmt.Sprintf("cookie_substr_key_value_b64_type:*%s -> %s", keySuffix, expect))
							case "string", "str":
								if utf8.Valid(b) {
									matchedBy = append(matchedBy, fmt.Sprintf("cookie_substr_key_value_b64_type:*%s -> %s", keySuffix, expect))
								}
							}
							break
						}
					}
				}

			default:
				// ignore unknown type
			}
		}
		if len(matchedBy) > 0 {
			results = append(results, matchResult{Name: c.Name, MatchedBy: matchedBy})
		}
	}
	return results
}

func main() {
	var (
		host     string
		raw      bool
		jsonMode bool
		timeout  int
		insecure bool
		ua       string
	)
	flag.StringVar(&host, "host", "", "Host or URL to scan (e.g., example.com or https://example.com)")
	flag.BoolVar(&raw, "raw", false, "Print only result name(s) in lowercase with underscores; print 'null' on no match.")
	flag.BoolVar(&jsonMode, "json", false, "Output a structured JSON result.")
	flag.IntVar(&timeout, "timeout", 10, "HTTP timeout in seconds (default: 10).")
	flag.BoolVar(&insecure, "insecure", false, "Skip TLS verification (verify=false).")
	flag.StringVar(&ua, "ua", "", "Custom User-Agent string.")
	flag.Parse()

	if host == "" && !(jsonMode || raw) {
		fmt.Print(PURPLE + "[?] Enter host to scan : " + END)
		reader := bufio.NewReader(os.Stdin)
		line, _ := reader.ReadString('\n')
		host = strings.TrimSpace(line)
	}
	if host == "" {
		if jsonMode {
			j := jsonOut{Error: "Missing --host"}
			enc := json.NewEncoder(os.Stdout)
			enc.SetEscapeHTML(false)
			_ = enc.Encode(j)
		} else {
			fmt.Println(RED + "[!] Missing --host" + END)
		}
		os.Exit(1)
	}

	host = normalizeHost(host)

	// Load fingerprints
	rawFP, err := readFingerprints()
	if err != nil {
		if jsonMode {
			j := jsonOut{Host: host, Error: fmt.Sprintf("Could not parse fingerprints: %v", err)}
			enc := json.NewEncoder(os.Stdout)
			enc.SetEscapeHTML(false)
			_ = enc.Encode(j)
		} else {
			fmt.Println(RED + fmt.Sprintf("[!] Could not parse fingerprints: %v", err) + END)
		}
		os.Exit(1)
	}
	prepared := prepare(rawFP)

	if !(raw || jsonMode) {
		fmt.Println(PURPLE + BOLD + "CMS Detector" + END)
		fmt.Println(PURPLE + "[•] Made by: https://github.com/joshuavanderpoll/CMS-Detector" + END)
		fmt.Printf(BLUE+"[@] Scanning host "+DARKCYAN+"\"%s\""+BLUE+"..."+END+"\n", host)
	}

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: insecure}, //nolint:gosec
	}
	client := &http.Client{
		Timeout:   time.Duration(timeout) * time.Second,
		Transport: transport,
	}
	// track redirects
	redirects := 0
	client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
		redirects = len(via)
		return nil
	}

	req, err := http.NewRequest("GET", host, nil)
	if err != nil {
		printErrAndExit(host, err, jsonMode)
	}
	if ua == "" {
		ua = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36"
	}
	req.Header.Set("User-Agent", ua)
	req.Header.Set("Cache-Control", "max-age=0")
	req.Header.Set("sec-ch-ua", "\"Chromium\";v=\"126\", \"Google Chrome\";v=\"126\", \"Not;A=Brand\";v=\"99\"")
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", "\"macOS\"")
	req.Header.Set("sec-fetch-dest", "document")
	req.Header.Set("sec-fetch-mode", "navigate")
	req.Header.Set("sec-fetch-site", "none")
	req.Header.Set("upgrade-insecure-requests", "1")
	req.Header.Set("referer", host)

	start := time.Now()
	resp, err := client.Do(req)
	elapsed := time.Since(start)
	if err != nil {
		printErrAndExit(host, err, jsonMode)
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)

	// match on lowercased, ' -> " replaced, trimmed
	bodyLower := strings.ToLower(strings.ReplaceAll(string(b), "'", "\""))
	bodyLower = strings.TrimSpace(bodyLower)

	matches := match(prepared, resp, bodyLower)
	detected := len(matches) > 0

	if jsonMode {
		out := jsonOut{
			Host:       host,
			StatusCode: resp.StatusCode,
			Detected:   detected,
			Matches:    matches,
			TimingMS:   elapsed.Milliseconds(),
			Redirects:  redirects,
		}
		enc := json.NewEncoder(os.Stdout)
		enc.SetEscapeHTML(false)
		_ = enc.Encode(out)
	} else if raw {
		if detected {
			for _, m := range matches {
				fmt.Println(GREEN + strings.ReplaceAll(strings.ToLower(strings.ReplaceAll(m.Name, " ", "_")), "’", "'") + END)
			}
		} else {
			fmt.Println("null")
		}
	} else {
		if detected {
			names := make([]string, 0, len(matches))
			for _, m := range matches {
				names = append(names, fmt.Sprintf("\"%s\"", m.Name))
			}
			fmt.Printf(GREEN+"[√] \"%s\" is using "+BLUE+"%s"+GREEN+"!"+END+"\n", host, strings.Join(names, ", "))
			for _, m := range matches {
				reasons := m.MatchedBy
				if len(reasons) > 3 {
					reasons = append(reasons[:3], fmt.Sprintf("(+%d more)", len(m.MatchedBy)-3))
				}
				fmt.Println(CYAN + "    ↳ matched by: " + strings.Join(reasons, "; ") + END)
			}
		} else {
			fmt.Println(YELLOW + "[!] No CMS could be detected." + END)
		}
	}
	if detected {
		os.Exit(0)
	}
	os.Exit(2)
}

func printErrAndExit(host string, err error, jsonMode bool) {
	if jsonMode {
		j := jsonOut{Host: host, Error: err.Error()}
		enc := json.NewEncoder(os.Stdout)
		enc.SetEscapeHTML(false)
		_ = enc.Encode(j)
	} else {
		fmt.Println(RED + fmt.Sprintf("[!] Could not retrieve host. Error: %v", err) + END)
	}
	os.Exit(1)
}
